RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,823 @@
+
+### Generics
+
+A type parameter is assumed to have non-nullable constraints when declared "normally":
+
+```fsharp
+type C<'T>() = // 'T is non-nullable
+class end
+```
+
+A warning is given if a nullable reference type is used to parameterize another type `C` that  does not accept nullable reference types:
+
+```fsharp
+type C<'T>() = class end
+
+let c = C<string | null> // WARNING
+```
+
+#### Constraints
+
+Today, there are two relevant constraints in F# - `null` and `not struct`:
+
+```fsharp
+type C1<'T when 'T: null>() = class end
+type C2<'T when 'T: not struct>() = class end
+```
+
+Both are actually the same thing for interoperation purposes, as they emit `class` as a constraint. See [Unresolved questions](nullable-reference-types.md#unresolved-questions) for more.
+
+If an F# class `B` has nullability as a constraint, then an inheriting class `D` also inherits this constraint, as per existing inheritance rules:
+
+```fsharp
+type B<'T | null>() = class end
+type D<'T | null>()
+    inherit B<'T>
+```
+
+That is, nullabulity constraints propagate via inheritance.
+
+If an unconstrained type parameter comes from an older C# assembly, it is assumed to be nullable, and warnings are given when non-nullability is assumed. If it comes from F# and it does not have the `null` constraint, then it is assumed to be non-nullable. If it does, then it is assumed to be nullable.
+
+Additionally, C# will introduce the `class?` constraint. This syntax sugar will compile into the `class` constraint, but the class itself will be decorated with an attribute that tells consumers that the `'T` is a nullable reference type. This attribute will need to be respected by the F# compiler.
+
+### Type inference
+
+Nullability is propagated through type inference:
+
+```fsharp
+// Inferred signature:
+//
+// val makeNullIfEmpty : str:string -> string | null
+let makeNullIfEmpty (str: string) =
+    match str with
+    | "" -> null
+    | _ -> str
+
+ // val xs : (string | null) list
+let xs = [ ""; ""; null ]
+
+// val ys : (string | null) []
+let ys = [| ""; null; "" |]
+
+// val zs : seq<string | null>
+let zs = seq { yield ""; yield null }
+
+// val x : 'a | null
+let x = null
+
+// val s : string
+let mutable s = "hello"
+
+// s is now (s | null), despite the warning
+s <- null
+```
+
+This includes function composition:
+
+```fsharp
+// val f1 : 'a -> 'a
+let f1 s = s
+
+// val f2 : string -> (string | null)
+let f2 s = if s <> "" then "hello" else null
+
+// val f3 : string -> string
+let f3 (s: string) =
+    match s with
+    | null -> ""
+    | _ -> s
+
+// val pipeline : string -> (string | null)
+let pipeline = f1 >> f2 >> f3 // WARNING - flowing possible null as input to function that assumes a non-null input
```

Comment:
I expect nullness will be checked through standard type inference.  So the warning message would be more likely to be something akin to "warning: nullness mismatch.  The type ... did not match the type ...`